src/builder.rs

```diff
@@ -53,71 +50,102 @@ pub fn run(root_nix_file: &NixFile, cas: &ContentAddressable) -> Result<Info, Er
 
     debug!("$ {:?}", cmd);
 
-    let mut child = cmd.spawn()?;
-
-    let stdout = child
-        .stdout
-        .take()
-        .expect("we must be able to access the stdout of nix-build");
-    let stderr = child
-        .stderr
-        .take()
-        .expect("we must be able to access the stderr of nix-build");
-
-    let stderr_results: thread::JoinHandle<Vec<LogDatum>> = thread::spawn(move || {
-        let reader = BufReader::new(stderr);
-        reader
-            .lines()
-            .map(|line| parse_evaluation_line(&line.unwrap()))
-            .collect::<Vec<LogDatum>>()
-    });
-
-    let produced_drvs: thread::JoinHandle<Vec<PathBuf>> = thread::spawn(move || {
-        BufReader::new(stdout)
-            .lines()
-            .map(|line| PathBuf::from(line.unwrap()))
-            .collect::<Vec<PathBuf>>()
-    });
-
-    let (exec_result, drvs, results) =
-        (child.wait()?, produced_drvs.join()?, stderr_results.join()?);
-
-    let (paths, named_drvs, log_lines): (Vec<PathBuf>, HashMap<String, PathBuf>, Vec<String>) =
-        results.into_iter().fold(
-            (vec![], HashMap::new(), vec![]),
-            |(mut paths, mut named_drvs, mut log_lines), result| {
+    let output = cmd.spawn()?.wait_with_output()?;
+
+    let stderr_results =
+        ::nix::parse_nix_output(&output.stderr, |line| parse_evaluation_line(line));
+
+    let produced_drvs = ::nix::parse_nix_output(&output.stdout, StorePath::from);
+
+    // iterate over all lines, parsing out the ones we are interested in
+    let (paths, output_paths, log_lines): (
+        Vec<PathBuf>,
+        // `None` if the field was not seen before, `Some` if it was
+        OutputPaths<Option<DrvFile>>,
+        Vec<OsString>
+    ) =
+    stderr_results.into_iter().fold(
+        (vec![], OutputPaths { shell: None, shell_gc_root: None }, vec![]),
+        |(mut paths, mut output_paths, mut log_lines), result| {
                 match result {
                     LogDatum::Source(src) => {
                         paths.push(src);
                     }
-                    LogDatum::AttrDrv(name, drv) => {
-                        named_drvs.insert(name, drv);
-                    }
+                    LogDatum::ShellDrv(drv) => {
+                        // check whether we have seen this field before
+                        match output_paths.shell {
+                            None => { output_paths.shell = Some(DrvFile(drv)); }
+                            // programming error
+                            Some(DrvFile(old)) => panic!(
+                                "`lorri read` got attribute `{}` a second time, first path was {:?} and second {:?}",
```

Comment:
Is this possible?

Reply:
Nope, that’s why it’s a programming error if it happens.

Reply:
outdated by rebase